module/VuFind/src/VuFind/Auth/Shibboleth.php

```diff
@@ -228,18 +250,12 @@ public function getSessionInitiator($target)
      */
     public function isExpired()
     {
-        $config = $this->getConfig();
-        if (isset($config->Shibboleth->username)
-            && isset($config->Shibboleth->logout)
-        ) {
-            // It would be more proper to call getServer on a Laminas request
-            // object... except that the request object doesn't exist yet when
-            // this routine gets called.
-            $username = isset($_SERVER[$config->Shibboleth->username])
-                ? $_SERVER[$config->Shibboleth->username] : null;
-            return empty($username);
-        }
-        return false;
+        $proxy = $this->getConfig()->Shibboleth->proxy ?? false;
+        // It would be more proper to call getServer on a Laminas request
+        // object... except that the request object doesn't exist yet when
+        // this routine gets called.
+        return ($proxy) ? !isset($_SERVER[$this->normalize(self::SHIB_SESSION_ID)])
+            : !isset($_ENV[self::SHIB_SESSION_ID]);
```

Comment:
I think `$_SERVER` and `$_ENV` are inverted here -- if proxy is true, we should look at the environment rather than the server, correct? Also, isn't normalization of keys normally associated with the environment, not the server?

Reply:
It is correct, but I noticed, that the logging of attributes is not working in proxy mode - fixed in the latest commit. See the comments in config.ini for Shibboleth proxy:

Set to true when shibboleth is running on the proxy side and so we have to read shibboleth attributes from headers instead of server environment variables

Are you running Shibboleth in the web server along with the Vufind or do you use reverse proxy in your environment?

Reply:
I'm running VuFind and Shibboleth on the same server, so I'm using `$_SERVER` rather than headers for attributes. Thus, my proxy variable is set to false. However, as written, this code appears to be saying "if proxy is true, then look at $_SERVER; else look at $_ENV... and that seems to be backwards to me).

Reply:
And do you use Apache web server with Shibboleth module?

Reply:
Yes, correct. So when I have an active Shibboleth session, `$_SERVER['Shib-Session-ID']` is set. When I do not, the value is unset. However, as written, this code is currently checking for `$_ENV['Shib-Session-ID']` when proxy is set to false, and for `$_SERVER['Shib_Session_ID']` (with - changed to _) when proxy is set to true, so it fails for me in both situations, because neither of those values are ever set. I still think this looks like a bug, and that the code should be 

<pre>
return ($proxy) ? !isset($_ENV[$this->normalize(self::SHIB_SESSION_ID)])
    : !isset($_SERVER[self::SHIB_SESSION_ID]);
</pre>

...but I apologize if I'm persisting in an incorrect assumption. :-)

Reply:
I also think it may be possible to use the request object instead of the superglobals here, because as far as I can tell, this method only gets through the manager's `checkForExpiredCredentials()` method, which in turn only gets called from this factory: https://github.com/vufind-org/vufind/blob/dc4677e100ecd786e5bb4b5b434a6ea08025aa9c/module/VuFind/src/VuFind/Auth/ManagerFactory.php#L95 -- I don't see why we couldn't add a `$container->get('Request')` parameter to this method call in this context).

Reply:
That is strange. Maybe you are using headers instead of environment variables for passing shibboleth attributes in your Apache configuration, which is not recommended:

ShibUseHeaders On

See https://wiki.shibboleth.net/confluence/display/SHIB2/NativeSPApacheConfig#NativeSPApacheConfig-AuthConfigOptions for details.

Reply:
Thanks for your latest adjustments to the code; they look likely to fix my problem. I will test when time permits (hopefully next week if not sooner).

Reply:
@xrosecky, I see that @EreMaijala has raised some new comments, so I'm going to hold off on testing until you have had a chance to discuss those with him. Please let me know when you're ready for me to test again!